homeassistant/components/hue/remote.py

```diff
@@ -0,0 +1,232 @@
+
+# state mapping for remote presses
+FOH_BUTTONS = {
+    16: "left_upper_press",
+    20: "left_upper_release",
+    17: "left_lower_press",
+    21: "left_lower_release",
+    18: "right_lower_press",
+    22: "right_lower_release",
+    19: "right_upper_press",
+    23: "right_upper_release",
+    100: "double_upper_press",
+    101: "double_upper_release",
+    98: "double_lower_press",
+    99: "double_lower_release",
+}
+RWL_BUTTONS = {
+    ZLL_SWITCH_BUTTON_1_INITIAL_PRESS: "1_click",
+    ZLL_SWITCH_BUTTON_2_INITIAL_PRESS: "2_click",
+    ZLL_SWITCH_BUTTON_3_INITIAL_PRESS: "3_click",
+    ZLL_SWITCH_BUTTON_4_INITIAL_PRESS: "4_click",
+    ZLL_SWITCH_BUTTON_1_HOLD: "1_hold",
+    ZLL_SWITCH_BUTTON_2_HOLD: "2_hold",
+    ZLL_SWITCH_BUTTON_3_HOLD: "3_hold",
+    ZLL_SWITCH_BUTTON_4_HOLD: "4_hold",
+    ZLL_SWITCH_BUTTON_1_SHORT_RELEASED: "1_click_up",
+    ZLL_SWITCH_BUTTON_2_SHORT_RELEASED: "2_click_up",
+    ZLL_SWITCH_BUTTON_3_SHORT_RELEASED: "3_click_up",
+    ZLL_SWITCH_BUTTON_4_SHORT_RELEASED: "4_click_up",
+    ZLL_SWITCH_BUTTON_1_LONG_RELEASED: "1_hold_up",
+    ZLL_SWITCH_BUTTON_2_LONG_RELEASED: "2_hold_up",
+    ZLL_SWITCH_BUTTON_3_LONG_RELEASED: "3_hold_up",
+    ZLL_SWITCH_BUTTON_4_LONG_RELEASED: "4_hold_up",
+}
+TAP_BUTTONS = {
+    ZGP_SWITCH_BUTTON_1: "1_click",
+    ZGP_SWITCH_BUTTON_2: "2_click",
+    ZGP_SWITCH_BUTTON_3: "3_click",
+    ZGP_SWITCH_BUTTON_4: "4_click",
+}
+Z3_BUTTON = {
+    1000: "initial_press",
+    1001: "repeat",
+    1002: "short_release",
+    1003: "long_release",
+}
+Z3_DIAL = {1: "begin", 2: "end"}
+
+
+async def async_setup_entry(hass, config_entry: ConfigEntry, async_add_entities):
+    """Defer remote setup to the shared sensor module."""
+    await hass.data[HUE_DOMAIN][
+        config_entry.entry_id
+    ].sensor_manager.async_register_component("remote", async_add_entities)
+
+
+class HueGenericRemote(GenericHueSensor, RemoteDevice):
+    """Parent class to hold common Hue Remote entity info."""
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return REMOTE_ICONS.get(self.sensor.modelid[0:3])
+
+    def turn_on(self, **kwargs):
+        """Do nothing."""
+
+    def turn_off(self, **kwargs):
+        """Do nothing."""
+
+    def _parsed_sensor_last_updated(self):
+        """
+        Localize the 'lastupdated' timestamp attribute from sensor data.
+
+        Adding the bridge timestamp to state attributes ensures that at least
+        the last remote press is processed as a state change.
+        """
+        return dt_util.parse_datetime(self.sensor.lastupdated).astimezone(
+            self.hass.config.time_zone
+        )
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        attributes = {
+            "model": self.sensor.type,
+            "last_button_event": self.state or "No data",
+            "last_updated": self._parsed_sensor_last_updated(),
```

Comment:
State in the state machine already has a last updated.

Reply:
Was my thought when I saw it, but this is another timestamp, also probably badly named. This is the real 'last_changed' stored in the bridge. It is not the same as our 'last_changed' or last_update, as they both refer to the state machine, not the sensor.

It is also used to force a state change event which would not appear otherwise when same button is pressed twice, I think (again, not sure as this is _inherit_ code)